[Witness Generation] Add witness query system & witness text generation

GlitchlessCode opened the pull request:
# Commit Notes
+ Add new `witness_gen` module
+ Add logic to run witness queries
+ Witness query data is unused as of right now

# Comments
This took longer than I thought it would. Not the programming itself but life has not been on the side of my time lately.

Either way, this adds a new `witness_gen` module which I plan to use for a lot more of the witness generation system. Here, I just wanted it to separate things out a little bit. Currently, I'm not doing anything with the query result, and in fact, I haven't had the opportunity to test if it even does *anything* but I'll try to quickly test that next, I'd like to know it's working before going on to generating the actual witnesses from this data. Also, witness queries that fail for some reason or another currently keep their own errors, I'd rather design the system to allow as many witnesses to be generated as possible, and then simply report error/gather diagnostics for any query or witness that fails. Unlike the lint queries themselves, it arguably should not fail hard, and should allow the rest of `c-s-c` to complete, it should just report any errors that occur to the user, and prompt them to report it as an issue. At least, that's my opinion on it, thoughts?

Additionally, as I state in one of the comments, if no witness query exists, we ideally want to forward existing lint_results, so those values can still be used in a witness that may not need a query. Though looking at this now, it might be optimal to add another case to the `match` for specifically where there *is* a witness to generate and *not* a witness query present to do that, and then on the catchall, instead just get filtered out (switching to filter_map instead of just map would allow this). However, this does still work, it's just a possible improvement for later.

obi1kenobi commented on src/witness_gen.rs:
```diff
@@ -0,0 +1,48 @@
+    lint_results: &Vec<BTreeMap<Arc<str>, FieldValue>>,
+) -> Vec<anyhow::Result<BTreeMap<Arc<str>, FieldValue>>> {
+    lint_results
+        .clone()
+        .into_iter()
```
Consideer `.iter().cloned()` instead since it's more efficient: it only clones the elements without also needing to allocate a whole `Vec` which we will immediately discard.

GlitchlessCode replied:
Fixing that right now, good point, thanks!

obi1kenobi commented on src/witness_gen.rs:
```diff
@@ -0,0 +1,48 @@
+            let witness_results = adapter
+                .run_query(&witness_query.query, arguments)
+                .and_then(|mut query_results| {
+                    let query_result = query_results.next();
+
+                    match query_results.next() {
```
In principle this should be fine, but it's probably a bit better to play it safe since it won't cost us anything.

The issue is that not all `impl Iterator` types allow calling `.next()` on them after one `.next()` has already returned `None`. The `Iterator` trait doesn't specify what happens — they might return `None` again, they might `panic!()`, they might return `Some(..)` etc.

An `impl FusedIterator` resolves this and guarantees that once one `.next()` call returns `None`, all subsequent `.next()` calls also do the same.

But here we can just as easily check if the first `.next()` was `None`, raise the error, and avoid the second call altogether.

obi1kenobi commented on src/witness_gen.rs:
```diff
@@ -0,0 +1,48 @@
+pub(crate) fn run_witness_queries(
+    adapter: &VersionedRustdocAdapter,
+    witness_query: &WitnessQuery,
+    lint_results: &Vec<BTreeMap<Arc<str>, FieldValue>>,
+) -> Vec<anyhow::Result<BTreeMap<Arc<str>, FieldValue>>> {
```
Do we actually need this "`Vec` in `Vec` out" interface? I'm not sure what we gain by having this extra layer of abstraction, compared to just having a function that takes one lint result and produces the corresponding witness query output (or an error).

obi1kenobi commented on src/check_release.rs:
```diff
@@ -342,6 +344,34 @@ pub(super) fn run_check_release(
+                        semver_query,
+                        witness_gen::run_witness_queries(&adapter, witness_query, lint_results),
+                    ),
+                    // If no witness query exists, we still want to forward the existing output
+                    _ => (
+                        semver_query,
+                        lint_results
+                            .clone()
+                            .into_iter()
+                            .map(Result::Ok)
+                            .collect_vec(),
+                    ),
```
I'm confused by this case. If there's no witness data, why are we cloning `lint_results`? I don't see what we'd do with it.

GlitchlessCode replied:
This was the thing I was saying above. Technically it's not just if there's no witness data, just if there's no *query*. Either way, fixing this.

obi1kenobi reviewed (commented):
> Also, witness queries that fail for some reason or another currently keep their own errors, I'd rather design the system to allow as many witnesses to be generated as possible, and then simply report error/gather diagnostics for any query or witness that fails.

This seems reasonable.

> Unlike the lint queries themselves, it arguably should not fail hard, and should allow the rest of `c-s-c` to complete, it should just report any errors that occur to the user, and prompt them to report it as an issue. At least, that's my opinion on it, thoughts?

Also reasonable! Let's do that.

> Additionally, as I state in one of the comments, if no witness query exists, we ideally want to forward existing lint_results, so those values can still be used in a witness that may not need a query. Though looking at this now, it might be optimal to add another case to the `match` for specifically where there _is_ a witness to generate and _not_ a witness query present to do that

I like the latter idea better! The former would be confusing and probably a bit convoluted.

> and then on the catchall, instead just get filtered out (switching to filter_map instead of just map would allow this).

I'm afraid I didn't follow this part.

Overall, I think it'd be useful to have a bit more functionality fleshed out here before we merge this, because for all we know some of the pieces might not quite fit well enough together. Better to find that out ASAP :)

obi1kenobi commented:
https://github.com/obi1kenobi/cargo-semver-checks/pull/1343 will fix the beta tests, and I believe nightly as well.

GlitchlessCode commented:
# Commit Notes
 Break out map, fix small issues

+ Extracts the mapping logic for witness queries to their own helper function above `run_check_release`
+ Switches mapping logic to use `filter_map`, allowing dropping of unused lint results
+ Removes unecessary vector collection, maintaining iterator format of witness query results
+ Switch to faster `.iter().cloned()` rather than `.clone().into_iterator()`
+ Fix potential issue with iterator in `run_witness_query`, returning early if `None` found

obi1kenobi commented on src/witness_gen.rs:
```diff
@@ -1,48 +1,43 @@
+            } else {
+                // If there is no query match, something has gone very wrong
+                Err(anyhow::anyhow!(
+                    "witness query should match exactly one time, matched zero times"
```
Imagine debugging a problem where this branch is hit.

You notice a `cargo-semver-checks` invocation where this error happens. You see: `witness query should match exactly one time, matched zero times` as the output.

What do you do next?

obi1kenobi commented on src/check_release.rs:
```diff
@@ -230,6 +231,46 @@ fn print_triggered_lint(
+    adapter: &VersionedRustdocAdapter,
+) -> Option<(
+    &'query SemverQuery,
+    Vec<anyhow::Result<BTreeMap<Arc<str>, FieldValue>>>,
```
I'm not entirely sold on returning this `Vec` as opposed to immediately acting on the results.

Consider writing out more of the witness logic (expanding beyond the original scope of the PR) so we can see what fits best?

GlitchlessCode commented:
# Commit notes

Add witness text generation, refactor witness system

+ Add `QueryResult` type for simplicity
+ Add `run_witness_checks` function as witness generation entrypoint
+ Add error context
+ Add witness text generation immediately after witness query running to avoid passing witness query data around
+ TODO: Snapshot witness query data
+ Remove witness running logic from `check_release.rs`, replacing with single entrypoint call

# Comments
Took what we discussed about adding more context to the errors, added them as `anyhow` context. Also removed the passing around of vectors of witness query result data, instead immediately generating witness text from handlebars templates. Assuming all is well here, and there may be more work necessary, but I'm pretty happy with it, next step is to take the witness text and actually generate all the witness crates using the computed witness text.

obi1kenobi commented on src/witness_gen.rs:
```diff
@@ -0,0 +1,148 @@
+            }
+        })
+        .with_context(|| {
+            format!("Error running witness query with input arguments {arguments:?}")
```
In the common case, we don't hit this error. So let's not make the common case clone `arguments` just for the uncommon case of this error to be able to print them. We can reconstruct `arguments` if we need it here.

GlitchlessCode replied:
Could you go more in depth for this one? I get what we don't want, I don't like that clone either, I just wasn't sure how to get rid of it.

GlitchlessCode replied:
Oh I see, yeah that makes sense! Thanks Predrag

obi1kenobi commented on src/witness_gen.rs:
```diff
@@ -0,0 +1,148 @@
+                    let witness_results = run_witness_query(adapter, witness_query, lint_result)
+                        .with_context(|| {
+                            format!(
+                                "Error running witness query for {}: {}",
+                                semver_query.id, semver_query.human_readable_name
```
```suggestion
                                "Error running witness query for {}",
                                semver_query.id
```
Since our human-readable names aren't very good, and since users generally interact with lints via the `id` — for example when configuring them in `Cargo.toml`.

obi1kenobi commented on src/witness_gen.rs:
```diff
@@ -0,0 +1,148 @@
+                    generate_witness_text(handlebars, witness_template, witness_results)
+                        .with_context(|| {
+                            format!(
+                                "Error generating witness text for witness {}: {}",
+                                semver_query.id, semver_query.human_readable_name
```
```suggestion
                                "Error generating witness text for witness {}",
                                semver_query.id,
```

obi1kenobi commented on src/witness_gen.rs:
```diff
@@ -0,0 +1,148 @@
+                    generate_witness_text(handlebars, witness_template, lint_result).with_context(
+                        || {
+                            format!(
+                                "Error generating witness text for queryless witness {}: {}",
+                                semver_query.id, semver_query.human_readable_name
```
```suggestion
                                "Error generating witness text for queryless witness {}",
                                semver_query.id
```

obi1kenobi commented on src/witness_gen.rs:
```diff
@@ -0,0 +1,148 @@
+    adapter: &VersionedRustdocAdapter,
+    lint_results: &[(&SemverQuery, Duration, Vec<QueryResults>)],
+) {
+    // Have to pull out handlebars, since &GlobalConfig cannot be shared across threads
+    let handlebars = config.handlebars();
```
Oh, fascinating. I didn't know the type wasn't thread-safe. We might want to fix this at some point when we have no lower-hanging fruit to pick — but very nice comment in the meantime!

obi1kenobi reviewed (commented):
Good stuff! I think just one more polishing pass and we should be able to merge this.

obi1kenobi commented:
Btw I fixed the broken lint job, so merging the upstream `main` should fix your CI (except `nightly`).

GlitchlessCode commented on src/witness_gen.rs:
```diff
@@ -0,0 +1,148 @@
+fn run_witness_query(
+    adapter: &VersionedRustdocAdapter,
+    witness_query: &WitnessQuery,
+    mut lint_result: QueryResults,
+) -> Result<QueryResults> {
```
Good to know. Personally just had that in for reading ergonomics, but I'll go over that again.

GlitchlessCode commented on src/witness_gen.rs:
```diff
@@ -0,0 +1,148 @@
+/// results with the existing lint results. Each query must match exactly once, and will fail with an
+/// [`anyhow::Error`] otherwise.
+///
+/// Overlapping output keys between the [`WitnessQuery`] and the [`SemverQuery`](crate::query::SemverQuery)
+/// will result in the result from the [`WitnessQuery`] silently overriding the same key from the
+/// [`SemverQuery`](crate::query::SemverQuery).
```
Yeah I wasn't particularly happy with silent behaviour either. I'll get an error in there.

GlitchlessCode commented:
# Commit Notes
 Add error for overlap, and small fixes

+ Adds an error for when an output key from a witness query overlaps that of the lint query
+ Removes the QueryResults type alias, reverting back to full type information
+ Avoids cloning arguments on the most common code path
+ Removes human_readable_name use in errors

GlitchlessCode commented:
Clippy is complaining about a very long type and is recommending breaking something out to a `type`, which is exactly what I just reverted. Any thoughts on what to do?

obi1kenobi commented:
I would suppress it with `#[expect(clippy::type-complexity)]`. It means well and it's a valid complaint, but:
- this isn't our public API
- the complexity is inherent, not accidental
- we'll probably refactor this code more in the future.

obi1kenobi commented:
I quickly threw in the lint suppression and I'm marking this for merge as soon as CI finishes. Thanks for putting it together — onward! 🚀

obi1kenobi commented on src/witness_gen.rs:
```diff
@@ -0,0 +1,166 @@
+/// results with the existing lint results. Each query must match exactly once, and will fail with an
+/// [`anyhow::Error`] otherwise.
+///
+/// Overlapping output keys between the [`WitnessQuery`] and the [`SemverQuery`](crate::query::SemverQuery)
```
```suggestion
/// Overlapping output keys between the [`WitnessQuery`] and the [`SemverQuery`]
```
